### ISSUE TITLE
Development process

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#133]
Now that we ( @langner, @meatballs and I) all have push rights, should we agree to work on a `dev` branch that we have to ok before merging in to `master`? (Or does anyone have a better suggestion?).

This might be particularly relevant when the `packaging` branch gets merged in (hopefully soon :)).

**Comment by meatballs** [comment]
I was thinking about this last night and intended to raise something similar this morning!

I think there are two factors to consider:

1. Principle
    In my view, the principle should be that nothing is merged into master until it has been reviewed by somebody other than the original contributor.

    If stuff is coming in from others, then one of the three of us needs to ok it. If we are contributing ourselves, one of the other two needs to have given it the OK somehow.

2. Practicality
    Given that others will have to use the fork/pull request mechanism, I wonder if it's better to just use the same ourselves. For me, that has several advantages:

    * All contributions are visible in the same place
    * The process is consistent and simple to remember
    * The PR mechanism gives us a convenient place to comment on any contribution and, therefore, to review it and give it the OK

**Comment by drvinceknight** [comment]
<notifications@github.com> wrote:
>
> I was thinking about this last night and intended to raise something
similar this morning!
>
> I think there are two factors to consider:
>
> Principle
> In my view, the principle should be that nothing is merged into master
until it has been reviewed by somebody other than the original contributor.
>
> If stuff is coming in from others, then one of the three of us needs to
ok it. If we are contributing ourselves, one of the other two needs to have
given it the OK somehow.

Yup I think this is a good principle. Are we going to be super rigorous? (I
think it would be ok to push a typo fix directly for example...)

>
> Practicality
> Given that others will have to use the fork/pull request mechanism, I
wonder if it's better to just use the same ourselves. For me, that has
several advantages:
>
> All contributions are visible in the same place
> The process is consistent and simple to remember
> The PR mechanism gives us a convenient place to comment on any
contribution and, therefore, to review it and give it the OK

I think this is a good idea BUT I am currently terrified to fork this
repository as I have no idea what that will do to all the redirects since
the transfer (a fork would give me the same url right? Perhaps github will
be smart enough to fork it to a different url... Perhaps if I rename
immediately everything will not break... )...

I wonder if for us three we could just push to branches and open an issue
to discuss?

**Comment by drvinceknight** [comment]
(sorry about the formatting of above: on my phone - will fix).

**Comment by meatballs** [comment]
I still prefer the PR mechanism if there's a discussion required - it keeps those in one place and separate from 'genuine' issues - but then I already have a working fork!

(I believe that you should be able to fork a repo from an organisation to a personal account without any problem).

I suppose the question becomes, when is a discussion/review necessary and when isn't it? You're right in that a simple typo correction shouldn't need any review at all.

In the end, it's personal judgment, but I would probably ask myself the following:

* Am I certain that this work will not impact anything that anyone else is doing?
* Am I certain that what I've written works properly?
* Am I certain that what I've written was actually wanted/required by anyone else (I'd probably raise an issue before starting to answer that one).
* Am I confident that nobody is likely to have improvements they could make to what I've written?

**Comment by drvinceknight** [comment]
> (I believe that you should be able to fork a repo from an organisation to a personal account without any problem).

The only thing I wonder about that is that as soon as I fork it, the url of my project will be github.com/drvinceknight/Axelrod which is were various things are still pointing (and being redirected). Actually, screw it, let me have confidence that someone far smarter than me would have thought about this at github :)

I'm forking! :)

> In the end, it's personal judgment, but I would probably ask myself the following:

Re that, I expect that the 3 of us at least would be able to continue to communicate about these things and perhaps adjust as we go along? 

Let's go for the continual PR thing with 1 of us reviewing each other's work. So in essence, the status quo except for me who now has to commit via PR (which is cool and an improvement :)).

**Comment by langner** [comment]
I agree. In my experience, the best way to maintain code quality is to enforce review. The easiest way to do that on GitHub to require that somone other than the author merges each PR.

**Comment by meatballs** [comment]
How about something like this:

* If we're working on something by ourselves, we create a branch in our fork and push to it as we go along so that anyone can have a look at what we're up to. When we're done, we create a PR to master in this repo. Personally, I try to create an issue for anything I'm working and then name the branch for the issue number.
* If somebody else wishes to contribute to that work, they can create a PR against the branch in the fork.
* If we decide we're working in collaboration, we create a branch within this repo (again, I'd suggest using issue numbers for branch names) and submit PRs against that. When we're happy it's done, any of us can merge it into master.

**Comment by drvinceknight** [comment]
That sounds great to me.

**Comment by langner** [comment]
For me the review requirement is more than enough. Any branching mechanism that works is fine - I don't think it will matter where the commits are coming from.

**Comment by meatballs** [comment]
> For me the review requirement is more than enough. Any branching mechanism that works is fine - I don't think it will matter where the commits are coming from.

Wise words - I'm probably over-thinking this!

**Comment by drvinceknight** [comment]
I'm going to go ahead and close this.
